### PR TITLE
Implementing Gather OP

### PR DESCRIPTION
**Pull request by kevinch-nv**

### CODE REVIEW

**Comment by kevinch-nv** [comment]
Any update on this PR getting reviewed?

**Comment by yinghai** [comment]
@kevinch-nv How is the progress on setting up the CI?

**Comment by kevinch-nv** [comment]
@yinghai We are discussing setting up CI internally. Do you know of what kind of CI there is for ONNX as a whole? Are there existing testing infrastructure as well? We can perhaps build on top of what already exists.

**Comment by yinghai** [comment]
For now, we can just use `onnx_backend_test.py`. Once it is setup, we can add more from onnx backend test.

**Comment by yinghai** [comment]
> We are discussing setting up CI internally. 

Setting up CI internally may discourage OSS contributions. If the effort is similar, I'd suggest we setup CI on github.

**Comment by kevinch-nv** [comment]
> > We are discussing setting up CI internally.
> 
> Setting up CI internally may discourage OSS contributions. If the effort is similar, I'd suggest we setup CI on github.

What I meant to say is that we are discussing how we want to set up CI internally. The actual implementation will most likely be public.

**Comment by yinghai** [comment]
It's all good then! :)

File: builtin_op_importers.cpp [hunk 1]
```diff
@@ -877,6 +877,16 @@ DEFINE_BUILTIN_OP_IMPORTER(Flatten) {
   return {{tensor_ptr}};
 }
 
+#if NV_TENSORRT_MAJOR >= 4
+DEFINE_BUILTIN_OP_IMPORTER(Gather) {
+    nvinfer1::ITensor& data = convertToTensor(inputs.at(0), ctx);
+    nvinfer1::ITensor& indices = convertToTensor(inputs.at(1), ctx);
+    OnnxAttrs attrs(node);
+    int axis = attrs.get<int>("axis");
+    RETURN_FIRST_OUTPUT(ctx->network()->addGather(data, indices, axis));
```
[the remark this answers is not in the document]
Comment (kevinch-nv) [reply builtin_op_importers.cpp]: Good comments - addressed in latest commits.

File: builtin_op_importers.cpp [hunk 2]
```diff
@@ -882,7 +882,21 @@ DEFINE_BUILTIN_OP_IMPORTER(Gather) {
...
+      axis += ndim;
+    }
+
+    // TRT does not support operations across the batch dimension
+    if (inputs.at(0).is_tensor())
+    {
+      --axis; 
+    }
+    
+    ASSERT(0 <= axis && axis < ndim, ErrorCode::kINVALID_NODE);
```
Comment (kevinch-nv) [inline builtin_op_importers.cpp]: This can be simplified by the convert_axis utility function in https://github.com/onnx/onnx-tensorrt/pull/103.

**Pull request merged**